extractor/docker/docker.go

```diff
@@ -257,29 +311,138 @@ func (d Extractor) extractLayerFiles(layerCh chan layer, errCh chan error, ctx c
+		Value:      &cacheContent,
+	})
+
+	if found {
+		b, err := extractTarContent(cacheContent)
+		if err != nil || len(b) <= 0 {
+			found = false
+		}
+
+		_, err = tarContent.Write(b)
+		if err != nil {
+			found = false
+		}
+	}
+
+	if !found {
+		rc, err := r.DownloadLayer(ctx, image.Path, dig)
 		if err != nil {
 			errCh <- xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
 			return
 		}
+		defer rc.Close()
 
-		rc, err = d.Cache.Set(string(dig), layerRC)
+		// read the incoming gzip from the layer
+		gzipReader, err := gzip.NewReader(rc)
 		if err != nil {
-			log.Print(err)
+			errCh <- xerrors.Errorf("could not init gzip reader: %w", err)
+			return
+		}
+		defer gzipReader.Close()
+
+		tarReader := tar.NewReader(io.TeeReader(gzipReader, &tarContent))
+
+		var storeInCache bool
+		var cacheBuf bytes.Buffer
+		if len(filenames) > 0 {
+			if cacheBuf, storeInCache, err = getFilteredTarballBuffer(tarReader, filenames); err != nil {
+				errCh <- err
+				return
+			}
+		}
+
+		// only store in cache if a required file was found
+		// TODO: should we error out in case of a cache store failure?
+		if storeInCache {
+			d.storeLayerInCache(cacheBuf, dig)
 		}
 	}
-	gzipReader, err := gzip.NewReader(rc)
+
+	layerCh <- layer{ID: dig, Content: ioutil.NopCloser(bytes.NewReader(tarContent.Bytes()))}
+	return
+}
+
+func extractTarContent(cacheContent []byte) ([]byte, error) {
+	var tarContent []byte
+
+	dec, err := zstd.NewReader(nil)
 	if err != nil {
-		errCh <- xerrors.Errorf("invalid gzip: %w", err)
+		return nil, err
+	}
+
+	tarContent, err = dec.DecodeAll(cacheContent, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return tarContent, nil
+}
+
+func getFilteredTarballBuffer(tr *tar.Reader, requiredFilenames []string) (bytes.Buffer, bool, error) {
+	var requiredFileFound bool
+
+	var cacheBuf bytes.Buffer
+	// Create a new tar to store in the cache
+	twc := tar.NewWriter(&cacheBuf)
+	defer twc.Close()
+
+	// check what files are inside the tar
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // end of archive
+		}
+		if err != nil {
+			return cacheBuf, false, xerrors.Errorf("%s: invalid tar: %w", ErrFailedCacheWrite, err)
+		}
+		for _, file := range requiredFilenames {
+			if file == hdr.Name {
+				requiredFileFound = true
+
+				hdrtwc := &tar.Header{
+					Name: hdr.Name,
+					Mode: 0600,
+					Size: hdr.Size,
+				}
+
+				if err := twc.WriteHeader(hdrtwc); err != nil {
+					return cacheBuf, requiredFileFound, xerrors.Errorf("%s: %s", ErrFailedCacheWrite, err)
+				}
+
+				_, err := io.Copy(twc, tr)
+				if err != nil {
+					return cacheBuf, requiredFileFound, xerrors.Errorf("%s: %s", ErrFailedCacheWrite, err)
+				}
+				break
+			}
+		}
+	}
+	return cacheBuf, requiredFileFound, nil
+}
+
+func (d Extractor) storeLayerInCache(cacheBuf bytes.Buffer, dig digest.Digest) {
+	// compress tar to zstd before storing to cache
+	e, err := zstd.NewWriter(nil, zstd.WithEncoderLevel(zstd.SpeedDefault))
```

Comment:
cacheBuf is a bit large, so it may be better to use io.Copy.
https://github.com/klauspost/compress/tree/master/zstd#usage

>Smaller encodes are encouraged to use the EncodeAll function.

```suggestion
	var output bytes.Buffer
	e, err := zstd.NewWriter(output, zstd.WithEncoderLevel(zstd.SpeedDefault))
```